MSC3174: An error code for spam rejections (#3174)
matrix-org/matrix-doc: proposals/3174-an-error-code-for-spam-rejections.md

> # MSC3174: An error code for spam rejections

Reviewer: ```suggestion
# MSC3174: Error codes for spam rejections
```

Reviewer 2: There's only one new `errcode`, `M_ANTISPAM_REJECTION`

Reviewer: `kind`s are also error codes

Reviewer: just a separate error code from `errcode`

Some Matrix server implementations, including Synapse, support spam-check filters.
This proposal simply attempts to standardize spam rejection.

## Proposal

### Introduction

A spam-checker may decide to reject a message, a registration attempt, etc. for any
number of reasons. However, while every other rejection from the server has an
error code `M_*`, spam rejections do not have, rather hardcoding a message from the
server.

This has several consequences:

- error messages are hardcoded on the server and cannot be internationalized;
- there is no support for spam-checkers displaying different kinds of rejections
(e.g. "this account is temporarily deactivated pending investigation",
"registrations are temporarily deactivated due to suspicion of DoS",
"this server rejects federation from that server", etc.)

This proposal simply attempts to standardize HTTP responses to spam-check rejections.

The expection is that Matrix servers can adapt their API to take advantage of this
proposal.

### Proposal

1. Add a new error code `M_ANTISPAM_REJECTION`;
2. Any `M_ANTISPAM_REJECTION` error may come with a field `kind`, which may contain any of the following values:

>     - `m.reject.spam` -- spam has been detected (e.g. message or username is spam);
>     - `m.reject.inappropriate` -- unacceptable content has been detected (e.g. a user attempting to register `@deathtojews:...`);
>     - `m.reject.investigation` -- account is suspected of being a spambot and is deactivated pending investigation;
>     - `m.reject.server` -- federation with this server has been deactivated (e.g. because server is used to send spam);

Reviewer: ```suggestion
    - `m.reject.spam` -- spam has been detected (e.g. message or username is spam);
    - `m.reject.inappropriate` -- unacceptable content has been detected (e.g. a user attempting to register `@deathtojews:...`);
    - `m.reject.investigation` -- account is suspected of being a spambot and is deactivated pending investigation;
    - `m.reject.server` -- federation with this server has been deactivated (e.g. because server is used to send spam);
    - `m.reject.ua` --- disallowed user agent or user agent family (including implicitly detected behavioral patterns)
```

Reviewer 3: Do you mean bots that use specific useragents? That might be dangerous as plenty bots use default User-agents. especially the bad ones do. So I would think that it causes false positives.

Reviewer: > So I would think that it causes false positives.

Indeed it can. But nowhere in the proposed change tells the user agent detection must, or even should, be based on the `User-Agent` strings alone. Instead, the kinds of user agents that act similarly to the suspected user agent may be rejected by this error code.

Reviewer 2: If a server doesn't want to tell the user the reason, it shouldn't include a `kind` field at all.

Author: 1. The idea is that `kind` is optional, so as @tulir mentioned, if we don't want to tell the reason, just don't include a `kind`.
2. For `m.reject.investigation`, it doesn't have to be *illegal* activity. The notion of "illegal" is very complicated in Matrix-land, as each country has different laws. This is about any activity that requires investigation, typically spambots.

Author: While I don't think that `m.reject.ua` would cause any issue, I don't see a good use case for it yet. Unless there's a clear and present use case, I would suggest postponing `m.reject.ua` to another MSC.

# MSC3174: Error codes for spam rejections

Some Matrix server implementations, including Synapse, support spam-check filters.
This proposal simply attempts to standardize spam rejection.

## Proposal

### Introduction

A spam-checker may decide to reject a message, a registration attempt, etc. for any
number of reasons. However, while every other rejection from the server has an
error code `M_*`, spam rejections do not have, rather hardcoding a message from the
server.

This has several consequences:

- error messages are hardcoded on the server and cannot be internationalized;
- there is no support for spam-checkers displaying different kinds of rejections
(e.g. "this account is temporarily deactivated pending investigation",
"registrations are temporarily deactivated due to suspicion of DoS",
"this server rejects federation from that server", etc.)

This proposal simply attempts to standardize HTTP responses to spam-check rejections.

The expection is that Matrix servers can adapt their API to take advantage of this
proposal.

### Proposal

> 1. Add a new error code `M_ANTISPAM_REJECTION`;

Reviewer 4: Seeing as rejections could happen for reasons other than spam (for instance, `m.reject.server` does not explicitly specify spam to be the reason, only a possible reason), perhaps this should just be more generic, like `M_REJECTION`?

Author: The idea to call it `M_ANTISPAM_REJECTION` is due to the fact that the message is rejected by the antispam filter, even if it's not necessarily spam, but you're right, that's a fine nuance that doesn't bring anything to the table.

I'm ok with `M_REJECTION`.

> 2. Any `M_ANTISPAM_REJECTION` error MAY come with a field `kind`, which may contain any of the following values:

Reviewer: ```suggestion
2. Any `M_ANTISPAM_REJECTION` error MUST come with a field `kind`, which may contain any values, with the recommended set with corresponding below:
```
